src/auth/OAuth.ts

```diff
@@ -45,17 +45,211 @@ interface ClientCredentials {
+  /**
+   * The unique identifier of the target API you want to access.
+   */
+  audience?: string;
+  /**
+   * Resource Owner's identifier, such as a username or email address.
+   */
+  username: string;
+  /**
+   * Resource Owner's secret.
+   */
+  password: string;
+  /**
+   * String value of the different scopes the application is asking for. Multiple scopes are separated with whitespace.
+   */
+  scope?: string;
+  /**
+   * String value of the realm the user belongs. Set this if you want to add realm support at this grant.
+   * For more information on what realms are refer to https://auth0.com/docs/get-started/authentication-and-authorization-flow/resource-owner-password-flow#realm-support.
+   */
+  realm?: string;
+}
+
+export interface DeviceCodeGrantRequest {
+  /**
+   * Specify this to override the parent class's `clientId`
+   */
+  client_id?: string;
+
+  /**
+   * The device code previously returned from the `/oauth/device/code` endpoint.
+   */
+  device_code: string;
+}
+
+export interface RefreshTokenGrantRequest extends ClientCredentials {
+  /**
+   * Specify this to override the parent class's `clientId`
+   */
+  client_id?: string;
+
+  /**
+   * The Refresh Token to use.
+   */
+  refresh_token: string;
+
+  /**
+   * A space-delimited list of requested scope permissions.
+   * If not sent, the original scopes will be used; otherwise you can request a reduced set of scopes.
+   */
+  scope?: string;
+}
+
+export interface TokenExchangeGrantRequest {
+  /**
+   * Specify this to override the parent class's `clientId`
+   */
+  client_id?: string;
+
+  /**
+   * Externally-issued identity artifact, representing the user.
+   */
+  subject_token: string;
+
+  /**
+   * The unique identifier of the target API you want to access.
+   */
+  audience?: string;
+
+  /**
+   * String value of the different scopes the application is requesting.
+   * Multiple scopes are separated with whitespace.
+   */
+  scope?: string;
+
+  /**
+   * Optional element used for native iOS interactions for which profile updates can occur.
+   * Expected parameter value will be JSON in the form of: `{ name: { firstName: 'John', lastName: 'Smith }}`
+   */
+  user_profile: string;
+}
+
 /**
  *  OAuth 2.0 flows.
  */
 export default class OAuth extends BaseAuthAPI {
+  /**
+   * Perform an OAuth 2.0 grant.
+   * (You should only need this if you can't find the grant you need in this class.)
+   */
+  async grant<T = TokenSet>(
+    grantType: string,
+    bodyParameters: Record<string, any>,
+    initOverrides?: InitOverride
+  ): Promise<JSONApiResponse<T>> {
+    const response = await this.request(
+      {
+        path: '/oauth/token',
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/x-www-form-urlencoded',
+        },
+        body: new URLSearchParams({
+          client_id: this.clientId,
+          ...bodyParameters,
+          grant_type: grantType,
+        }),
```

Comment:
Just a question, if they need to use a grant that needs a secret, they should manually specify it ?

Reply:
Yep - we can't automatically add client credentials because not all grants need them (also this is a low level api that I wouldn't expect many to use - I'm only exposing it so we can use it in Passwordless)